t51-projects/gh-project-board.php

```diff
@@ -1,45 +1,70 @@
 <?php
 
 
-function curl_fetch() {
+$url = 'https://api.github.com/users/jessLundie'; // Set source URL
 
-	$url = 'https://api.github.com/users/jessLundie'; // Set source URL
+$media_header = array(
+	"Accept: application/vnd.github.inertia-preview+json",
+); // Custom media header required for projects in beta
 
-	$column_id = "15445284"; // Column id, found in the column link here: https://d.pr/i/W7RQJB
+$user_agent = "jessLundie"; // Set user agent
 
-	$project_url = "https://api.github.com/projects/columns/" . $column_id . "/cards"; // Set project URL
+$token = ""; // Basic Auth Token
 
-	$token = "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"; // Basic Auth Token
+$column_id = "15445284"; // Column id, found in the column link here: https://d.pr/i/W7RQJB
 
-	$user_agent = "jessLundie"; // Set user agent
+$project_url = "https://api.github.com/projects/columns/" . $column_id . "/cards"; // Set API URL for project
 
-	// Custom media header required for projects in beta
-	$headers = array(
-		"Accept: application/vnd.github.inertia-preview+json",
-	);
+function curl_fetch( $url, $media_header, $user_agent, $token, $api_url ) {
 
 	// Create a new cURL resource
 	$process = curl_init($url);
 
 	// Set options
-	curl_setopt( $process, CURLOPT_HTTPHEADER, $headers );
+	curl_setopt( $process, CURLOPT_HTTPHEADER, $media_header );
 	curl_setopt( $process, CURLOPT_USERAGENT, $user_agent );
 	curl_setopt( $process, CURLOPT_USERPWD, "$user_agent:$token" );
 	curl_setopt( $process, CURLOPT_RETURNTRANSFER, TRUE);
-	curl_setopt( $process, CURLOPT_URL, $project_url );
+	curl_setopt( $process, CURLOPT_URL, $api_url );
 
 	// Get project details
 	$result = curl_exec( $process );
 
 	// Close cURL
 	curl_close( $process );
 
+	$result = json_decode( $result, true ); // Convert response to array
 	return $result;
+
+}
+
+$cards = curl_fetch( $url, $media_header, $user_agent, $token, $project_url ); // List of cards
+
+// Create a list of all labels
+
+$label_list = [];
+
+foreach ( $cards as $card ) {
+
+	$label_url = $card['content_url'] . "/labels"; // Set API URL for labels
+
+	$label = curl_fetch( $url, $media_header, $user_agent, $token, $label_url  ); // API call to fetch labels
+
+	$label_list = array_merge( $label_list, $label ); // Add each label to list of labels
```

Comment:
although I applaud the functional way you are appending the labels to the list, we're in a kind of mixed territory here. we're swapping out the value of `$label_list` so we can't just know what its value is without trying to run the code in our heads, but we're picking an immutable joiner function whose performance is well behind the more normative _mutable_ code.

when creating a list like this in PHP by appending elements we normally use the `[] = ` syntax.

```php
$label_list[] = $label;
```

were we to want to use a functional/immutable approach we could, though it would be a little more awkward in PHP than it would be in JS.

```php
$label_list = array_reduce( $cards, function( $labels, $card ) {
    $label_url = …
    $label = curl_fetch( … );

    return array_merge( $labels, $label );
}, [] );
```

this version never _reassigns_ the value of `$label_list`. nonetheless, your PHPeers will likely prefer the `$label_list[] = $label` form

Reply:
Changed! But I'm not sure I understand the mutable / immutable distinction in this example. Can we discuss this more tomorrow?